WebKit is failing to register a service worker if script fetch is not fetched when document is being reloaded

youennf opened the pull request:
#### c53ee10740bf00f166a5b0e0b5daab3933bcd7e4
<pre>
WebKit is failing to register a service worker if script fetch is not fetched when document is being reloaded
<a href="https://rdar.apple.com/152398237">rdar://152398237</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294391">https://bugs.webkit.org/show_bug.cgi?id=294391</a>

Reviewed by Chris Dumez.

Some websites will test whether a page is controlled by a service worker or not.
If not, they will start registering a service worker and will reload until the page is controlled.

Reloading is expected to stop all loads, including the load to get the service worker script.
If the load is not finished by the time the reload cancels all loads, the service worker registration will fail.
This can make the website enter in a reload loop.

To prevent this, we add a quirk that delays reloading to happen until the service worker registration job has finished, if any.

* LayoutTests/http/wpt/service-workers/register-then-reload-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/register-then-reload.html: Added.
* LayoutTests/http/wpt/service-workers/resources/lengthy-worker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/resources/register-then-reload-iframe.html: Added.
* Source/WebCore/page/Location.cpp:
(WebCore::Location::reload):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDelayReloadWhenRegisteringServiceWorker const):
(WebCore::handleCapitalGroupQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::jobFailedWithException):
(WebCore::ServiceWorkerContainer::jobResolvedWithRegistration):
(WebCore::ServiceWorkerContainer::jobFinishedLoadingScript):
(WebCore::ServiceWorkerContainer::jobFailedLoadingScript):
(WebCore::ServiceWorkerContainer::hasServiceWorkerScriptFetch const):
(WebCore::ServiceWorkerContainer::whenScriptFetchJobFinished):
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerJob.cpp:
(WebCore::ServiceWorkerJob::fetchScriptWithContext):
(WebCore::ServiceWorkerJob::mayLoadScript const):
* Source/WebCore/workers/service/ServiceWorkerJob.h:

Canonical link: <a href="https://commits.webkit.org/296970@main">https://commits.webkit.org/296970@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d708dcdb5e0ade9e713d50ac6dd56a828e3d7ec9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116125 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60351 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38349 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83711 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113051 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99155 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64155 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23643 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17290 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59921 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93652 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17347 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37143 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92682 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37515 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95423 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92507 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23580 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37496 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15233 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33026 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42508 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36699 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38408 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2971,7 +2983,8 @@ void Quirks::determineRelevantQuirks()
         { "zoom"_s, &handleZoomQuirks },
-        { "dailymail"_s, &handleDailyMailCoUkQuirks }
+        { "dailymail"_s, &handleDailyMailCoUkQuirks },
+        { "capitalgroup"_s, &handleCapitalGroupQuirks }
```
I'm not the best reviewer for this, but it appears this list is meant to be in chronological order, but these last two entries break that. (Not sure if there's an easy way to assert that.)

youennf replied:
I moved it up.

youennf committed:
WebKit is failing to register a service worker if script fetch is not fetched when document is being reloaded
rdar://152398237
https://bugs.webkit.org/show_bug.cgi?id=294391

Reviewed by Chris Dumez.

Some websites will test whether a page is controlled by a service worker or not.
If not, they will start registering a service worker and will reload until the page is controlled.

Reloading is expected to stop all loads, including the load to get the service worker script.
If the load is not finished by the time the reload cancels all loads, the service worker registration will fail.
This can make the website enter in a reload loop.

To prevent this, we add a quirk that delays reloading to happen until the service worker registration job has finished, if any.

* LayoutTests/http/wpt/service-workers/register-then-reload-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/register-then-reload.html: Added.
* LayoutTests/http/wpt/service-workers/resources/lengthy-worker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/resources/register-then-reload-iframe.html: Added.
* Source/WebCore/page/Location.cpp:
(WebCore::Location::reload):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDelayReloadWhenRegisteringServiceWorker const):
(WebCore::handleCapitalGroupQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::jobFailedWithException):
(WebCore::ServiceWorkerContainer::jobResolvedWithRegistration):
(WebCore::ServiceWorkerContainer::jobFinishedLoadingScript):
(WebCore::ServiceWorkerContainer::jobFailedLoadingScript):
(WebCore::ServiceWorkerContainer::hasServiceWorkerScriptFetch const):
(WebCore::ServiceWorkerContainer::whenScriptFetchJobFinished):
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerJob.cpp:
(WebCore::ServiceWorkerJob::fetchScriptWithContext):
(WebCore::ServiceWorkerJob::mayLoadScript const):
* Source/WebCore/workers/service/ServiceWorkerJob.h:

Canonical link: https://commits.webkit.org/296970@main
